src/lib.rs

```diff
@@ -194,13 +194,59 @@ pub enum InterpolationType {
 /// A resampler that us used to resample a chunk of audio to a new sample rate.
 /// The rate can be adjusted as required.
 pub trait Resampler<T>: Send {
-    /// Resample a chunk of audio.
+    /// Resample a chunk of audio. 
+    ///
+    /// The input data is a slice, where each element of the slice is itself referenceable as a slice
+    /// ([`AsRef<[T]>`](AsRef)) which contains the samples for a single channel. Since [`Vec<T>`] implements
+    /// [`AsRef<[T]>`](AsRef), the input may simply be `&*Vec<Vec<T>>`. 
+    /// The output data is a vector, where each element of the vector is itself a vector
+    /// which contains the samples for a single channel.
+    /// TODO! The `active_channels_mask` is optional. (active_channels_mask: Option<&[bool]>)
+    /// Any channel marked as inactive by a false value will be skipped during processing
+    /// and the corresponding output will also be empty.
+    /// If None is given, the length of each input is used to determine the active channels.
+    /// Then if an input channel has zero length, this channel will be considered as inactive. 
+    //fn process<V: AsRef<[T]>>(&mut self, wave_in: &[V]) -> ResampleResult<Vec<Vec<T>>>;
+    fn process<V: AsRef<[T]>>(&mut self, wave_in: &[V]) -> ResampleResult<Vec<Vec<T>>> {
+        let mut wave_out = self.allocate_output_buffer();
+        let channel_mask = make_mask_from_buffers(wave_in);
+        self.process_into_buffer(wave_in, &mut wave_out, &channel_mask)?;
+        Ok(wave_out)
+    }
+
+    /// Resample a chunk of audio to a pre-allocated output buffer.
+    /// Use this in real-time applications where the unpredictable time required to allocate memory
+    /// is causing issues. If this is not a problem, then use the more convenient `process` method.
     ///
     /// The input data is a slice, where each element of the slice is itself referenceable as a slice
     /// ([`AsRef<[T]>`](AsRef)) which contains the samples for a single channel. Since [`Vec<T>`] implements
     /// [`AsRef<[T]>`](AsRef), the input may simply be `&*Vec<Vec<T>>`. The output data is a vector, where each element
     /// of the vector is itself a vector which contains the samples for a single channel.
-    fn process<V: AsRef<[T]>>(&mut self, wave_in: &[V]) -> ResampleResult<Vec<Vec<T>>>;
+    /// The output will be resized to fit all the output samples.
+    /// To avoid allocations, make sure that the output has sufficient capacity.
+    /// `active_channels_mask` is used to mark channels as active or inactive.
+    /// Any channel marked as inactive by a false value will be skipped during processing,
+    /// and the corresponding output will be left unchanged.
```

Comment:
I don't unnderstand the need for this. In what situation would a user want to pass an input Vec with some channels they don't want to resample?

Reply:
I do that in my DSP app. I create a resampler for all captured channels, then keep track of whether they are actually used later in the pipeline. For channels that aren't used (when muted for example) I skip resampling those. It's just to avoid unnecessary work. Creating a new resampler is expensive so I don't want to do that every time the active channels changes.

Reply:
Hmm, okay. I can't imagine a situation where I'd want to make use of that in my application. I like the idea to simplify the API by allowing to pass `None` instead of the user needing to pass a slice of bools.

Reply:
I started by making it optional but that gave too many "if let Some(mask) do this else do that".

Reply:
I want to try using this branch in my application but I'm not inclined to go through the trouble of doing that while there are still API changes planned.

Reply:
The only change I have planned is to the "process" method. I guess you will want to use the "process_to_buffer" and I don't think there will be any changes to that one. Only if I come up with a neat way of making the mask an Option.

Reply:
I would prefer if the mask was an Option for `process_into_buffer` as well. Otherwise it'll be up to my application to allocate another Vec for the mask and keep that associated with the buffer even though my application wouldn't have any use for it.

Reply:
It may help if I let None mean that all channels are used for `process_into_buffer`, without checking if they are empty. That should avoid at least some of the if-mess. I'll take a look.
The downside is that then the mask None option has different meaning in `process_into_buffer` and `process`. But I think that can be acceptable.

Reply:
> The downside is that then the mask None option has different meaning in process_into_buffer and process. But I think that can be acceptable.

:/ That could get confusing.

Reply:
I just pushed a version where they are as I described first. But I'll rework this so that `process_into_buffer` behaves the same as `process` and checks for empty channels when given a `None`.

Reply:
Done!